pkg/controllers/applicationfailover/common.go

```diff
@@ -0,0 +1,79 @@
+package applicationfailover
+
+import (
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+type workloadUnhealthyMap struct {
+	sync.RWMutex
+	// key is the resource type
+	// value is also a map. Its key is the cluster where the unhealthy workload resides.
+	// Its value is the time when the unhealthy state was first observed.
+	workloadUnhealthy map[keys.ClusterWideKey]map[string]metav1.Time
+}
+
+func newWorkloadUnhealthyMap() *workloadUnhealthyMap {
+	return &workloadUnhealthyMap{
+		workloadUnhealthy: make(map[keys.ClusterWideKey]map[string]metav1.Time),
+	}
+}
+
+func (m *workloadUnhealthyMap) delete(key keys.ClusterWideKey) {
+	m.Lock()
+	defer m.Unlock()
+	delete(m.workloadUnhealthy, key)
+}
+
+func (m *workloadUnhealthyMap) set(key keys.ClusterWideKey, unHealthyClusters map[string]metav1.Time) {
+	m.Lock()
+	defer m.Unlock()
+	m.workloadUnhealthy[key] = unHealthyClusters
+}
+
+func (m *workloadUnhealthyMap) get(key keys.ClusterWideKey) map[string]metav1.Time {
+	m.RLock()
+	defer m.RUnlock()
+	return m.workloadUnhealthy[key]
+}
+
+func detectFailure(aggregatedStatusItems []workv1alpha2.AggregatedStatusItem, gracefulEvictionTasks []workv1alpha2.GracefulEvictionTask,
```

Comment:
This function couples too many responsibilities:
1. manages the unhealthy map
2. filter clusters that are already in the eviction tasks
3. select clusters that need to be evicted after grace period

We should consider splitting it into small functions.

Reply:
Through my analyze, there is a dependence in it because `unhealthy` map is used to select clusters that need to be evicted. We need to maintain the map meanwhile when we check whether the cluster need to be evicted.